scripts/casefold_db.py

```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python
+# Copyright 2021 The Matrix.org Foundation C.I.C.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import argparse
+import json
+import os
+import sqlite3
+import sys
+from typing import Any, Dict, List, Tuple
+
+import signedjson.sign
+
+from sydent.sydent import Sydent, parse_config_file
+from sydent.util import json_decoder
+from sydent.util.emailutils import sendEmail
+from sydent.util.hash import sha256_and_url_safe_base64
+from tests.utils import ResolvingMemoryReactorClock
+
+
+def calculate_lookup_hash(sydent, address):
+    cur = sydent.db.cursor()
+    pepper_result = cur.execute("SELECT lookup_pepper from hashing_metadata")
+    pepper = pepper_result.fetchone()[0]
+    combo = "%s %s %s" % (address, "email", pepper)
+    lookup_hash = sha256_and_url_safe_base64(combo)
+    return lookup_hash
+
+
+def update_local_associations(
+    sydent, db: sqlite3.Connection, send_email: bool, dry_run: bool
+):
+    """Update the DB table local_threepid_associations so that all stored
+    emails are casefolded, and any duplicate mxid's associated with the
+    given email are deleted.
+
+    :return: None
+    """
+    cur = db.cursor()
+
+    res = cur.execute(
+        "SELECT address, mxid FROM local_threepid_associations WHERE medium = 'email'"
+        "ORDER BY ts DESC"
+    )
+
+    # a dict that associates an email address with correspoinding mxids and lookup hashes
+    associations: Dict[str, List[Tuple[str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, and add to
+    # associations dict
+    for address, mxid in res.fetchall():
+        casefold_address = address.casefold()
+
+        # rehash email since hashes are case-sensitive
+        lookup_hash = calculate_lookup_hash(sydent, casefold_address)
+
+        if casefold_address in associations:
+            associations[casefold_address].append((address, mxid, lookup_hash))
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
```

Comment:
Why a tuple? It could just be a list of strings.

Reply:
From my reading it seemed that `executemany` needs to be passed a list of iterables-in this case tuples. Passing it a list of strings causes it to blow up with an `incorrect number of binding supplied` error. This actually confounded me for quite awhile and passing it a list of tuples was the best solution I could find, so if there's a cleaner/more correct way I would love to learn :)

Reply:
Oh, yes that's true, my bad, sorry, carry on :facepalm: